flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/AbstractNonTimeUnboundedPrecedingOver.java

```diff
@@ -0,0 +1,714 @@
+     * elements an UPDATE_BEFORE and UPDATE_AFTER event is emitted based on the previous and newly
+     * aggregated values. Updates are skipped if the previously accumulated value is the same as the
+     * newly accumulated value to save on network bandwidth. Implementation differs for rows vs
+     * range.
+     *
+     * @param delRow
+     * @param out
+     * @throws Exception
+     */
+    abstract void removeFromSortedList(RowData delRow, Collector<RowData> out) throws Exception;
+
+    /**
+     * Helper method to re-accumulate the aggregated value for all ids without the id that will be
+     * removed.
+     *
+     * @param ids
+     * @param delRow
+     * @return the index position of the id that should be removed
+     * @throws Exception
+     */
+    int reAccumulateIdsAndGetRemoveIndex(List<Long> ids, RowData delRow) throws Exception {
+        int removeIndex = -1;
+        for (int j = 0; j < ids.size(); j++) {
+            RowData curValue = valueMapState.get(ids.get(j));
+            if (valueEqualiser.equals(curValue, delRow)) {
+                removeIndex = j;
+            } else {
+                aggFuncs.accumulate(curValue);
+            }
+        }
+        return removeIndex;
+    }
+
+    /**
+     * Remove the ith element from the sortedList if the list of ids is empty and return the ith
+     * location OR update the sorted list with the list of ids and return the next location of the
+     * sortedList.
+     *
+     * @param sortedList
+     * @param idx
+     * @param ids
+     * @param curSortKey
+     * @return
+     */
+    int removeIdFromSortedList(
+            List<Tuple2<RowData, List<Long>>> sortedList,
+            int idx,
+            List<Long> ids,
+            RowData curSortKey) {
+        if (ids.isEmpty()) {
+            sortedList.remove(idx);
+        } else {
+            // Update sortedList without the removed id
+            sortedList.set(idx, new Tuple2<RowData, List<Long>>(curSortKey, ids));
+            idx++;
+        }
+        return idx;
+    }
+
+    /**
+     * Helper method to find the index of the id to be removed from the list of ids.
+     *
+     * @param ids
+     * @param delRow
+     * @return
+     * @throws Exception
+     */
+    int findIndexOfIdToBeRemoved(List<Long> ids, RowData delRow) throws Exception {
+        for (int j = 0; j < ids.size(); j++) {
+            RowData curValue = valueMapState.get(ids.get(j));
+            if (valueEqualiser.equals(curValue, delRow)) {
+                return j;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Helper method to get the previous accumulator based on the index of the sortedList and the
+     * index of the list of ids.
+     *
+     * @param sortedList
+     * @param i
+     * @param j
+     * @return
+     * @throws Exception
+     */
+    RowData getPreviousAccumulator(List<Tuple2<RowData, List<Long>>> sortedList, int i, int j)
```

Comment:
Move this to the ROWS operator?
It's only used there.

Reply:
Moved